### DISCUSSION TITLE
Please create a small example for a auth_callout service created in C#

### DISCUSSION THREAD

**Discussion by ploef** [nats-io/nats-server#6308] [Q&A]
Hi,

I try to implement a auth_callout service created with c#. I have it working now except for the signing part. the auth_callout configuration has the xkey set (created using nsc generate nkey --curve). 
I did find an example created in go (https://github.com/ConnectEverything/nats-by-example/blob/main/examples/auth/callout/cli/service/main.go). In that example, on line 52, the keypair is created for the xkey (SXA......). The .NET package also has the possibility with the "FromSeed" method. This does work for the issuerKey, but not for the xkey (Invalid Public Prefix Byte)

**Comment by Fixylone** [comment]
We're facing a very similar issue for a critical project we're working on. The xkey handling is essential for our implementation of the auth_callout service in .NET,.
 
We really appreciate any help or suggestions from the community, as this functionality is central to the success of our project. If anyone has made progress on this or has insights into how to resolve this issue, we'd be very grateful for your input!

The nkeys.net library (https://github.com/nats-io/nkeys.net) is currently under development, and it appears that the xkeys implementation is still in progress. Does anyone have any insights on when this feature might be ready?

Thanks a lot in advance, and looking forward to hearing from you!

**Reply by mtmk** [reply]
please check [1.0.0-preview.2](https://www.nuget.org/packages/NATS.NKeys/1.0.0-preview.2). it's not a full implementation but afaiu it should cover what you need. still in preview for API stability really. please feel free to provide feed back on its repo https://github.com/nats-io/nkeys.net so we can move it to stable soon.

**Reply by Fixylone** [reply]
Thank you for the quick response. I realize now that my question wasn’t as precise as it could have been—I’m specifically interested in the encryption/decryption functionality. Here’s a snippet of my NATS service configuration:

```
authorization {
  auth_callout {
    ...
    xkey: XAMHJVPKHHPYZQQM2IVWXKJH36KDDZZMSJ32QKSQBUODFX4I4HARO4GL
  }
}
```

> (Note: The key shown above is dummy key generated for testing.)

I do receive encrypted data in my custom auth_callout service. However, I haven’t found any support in the nkeys.net library for xkey decryption of the payload—unlike the functionality provided by the nats-io/nkeys library (https://github.com/nats-io/nkeys). Am I missing something, or is xkey encryption/decryption not supported in nkeys.net? Everything else works like a charm.. :)

**Reply by mtmk** [reply]
thanks @Fixylone for the details. that is correct we don't yet support that. can you open an issue on https://github.com/nats-io/nkeys.net please so i can prioritize it? thank you!

**Reply by Fixylone** [reply]
Done. @mtmk Thank you once again! 
https://github.com/nats-io/nkeys.net/issues/6

**Comment by mtmk** [comment]
xkeys generation is implemented in https://www.nuget.org/packages/NATS.NKeys/1.0.0-preview.2 let me know if you're having issues or feel free to create an issue on its repo https://github.com/nats-io/nkeys.net

**Comment by aricart** [comment]
Also I would encourage you to consider using https://github.com/synadia-io/callout.go for the callout implementation or at least to take a look at it to understand some of the nuances and security checks

**Comment by mtmk** [comment]
* in progress https://github.com/nats-io/nkeys.net/pull/7
